### ISSUE TITLE
Auth failure on FreeBSD

### ISSUE THREAD

**Issue by codedmart** [rblaze/haskell-dbus#42]
I use this package in various Linux distros, and have no problems, but when using it in FreeBSD I get
```
*** Exception: SocketError {socketErrorMessage = "Authentication failed", socketErrorFatal = True, socketErrorAddress = Just (Address "unix:guid=2fb1d926d67497d679a521255e41b85e,path=/tmp/dbus-4vBp0kLUee")}
```
I am not entirely sure what is going on. If I add some print statements I can see:
```
"AUTH EXTERNAL 31303031\r\n" <- sent
"REJECTED EXTERNAL" <- returned
```
The dbus session address is set. I tried connection with packages in python, and go just to be sure. Both of those are able to connect just fine. I don't know go but did notice in the go code there is some calls to [UnixCredentials](https://golang.org/src/syscall/sockcmsg_linux.go). Have no idea if the Haskell code is doing that for sure, but doesn't seem like it from what I can tell. Any ideas why this is failing on FreeBSD?

**Comment by rblaze** [comment]
Sorry, I have no clue here, and I have no FreeBSD around. Could you `strace` a simple test program and see if some syscall is failing?

**Comment by codedmart** [comment]
Sure here is some more debugging output. The part I find interesting is this from the go [output](https://gist.github.com/codedmart/209f7c4103d18baa8ac7b28211148828#file-go-test-out-L442). Which is the initial null byte sent. The equvilant in the haskell [output](https://gist.github.com/codedmart/209f7c4103d18baa8ac7b28211148828#file-go-test-out-L442). In the go one there is this with the initial null byte send `...level=SOL_SOCKET,type=SCM_CREDS...` which is what I linked above with the UnixCredentials part. Thanks!

**Comment by codedmart** [comment]
I checked python as well. Seems it is sending something similar for the initial null byte:
```
sendmsg(3,{NULL,0,[{"\0",1}],1,{{level=SOL_SOCKET,type=SCM_CREDS,data={0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00}}},104,0},MSG_NOSIGNAL) = 1 (0x1)
```
I will keep digging into this.

**Comment by codedmart** [comment]
Well I was able to get this working, but it is pretty rough right now. I didn't see currently in the network package the ability to use `sendmsg` with `cmsg` which is needed to properly send credentials for the socket. There is a PR open to add this ability to network [here](https://github.com/haskell/network/pull/433). If I am reading things correctly it looks like this isn't needed in Linux because `abstract` is used rather than `path` for the unix socket which is why things work correctly there. I ended up using the `network-msg` package to get this working for me locally, but I am not sure adding that as another dep makes sense if the PR in network lands soon. So feel free to leave open or close and I will follow up when the PR in network lands.
